### PR TITLE
feat(gym): per-session detail page with HR sample stream (#165)

### LINKED ISSUE #165: feat(gym): per-session detail page with HR sample stream
## Summary

Add a per-session detail page (`/training-facility/gym/session/[started_at]`) that renders a full HR curve, zone breakdown, and per-session metadata for a single workout. Requires storing the raw HR sample stream Apple Watch produces during a workout, which `scripts/preprocess-health.py` currently aggregates into zone seconds and discards.

Single-user site — storage cost is bounded and well under the free tier, so the usual "scale considerations" objection doesn't apply. Greenlit to ship; no longer staged behind the sibling issues.

## Storage cost — fits free tier comfortably

For a single user (~4 sessions/week × 45 min × 12 samples/min at 5s Apple Watch sampling):

- ~2,200 samples/session × ~200 sessions/year = ~440k rows/year
- ~80–120 bytes/row (timestamptz + numeric + session FK + index overhead)
- **~50 MB/year** — Supabase free tier (500 MB) holds ~10 years

Storage is a non-issue. Engineering effort lives in the preprocessor + import-script changes; the page itself reuses existing chart primitives.

## Architecture

**New table** (migration following #152's pattern):

```sql
create table public.cardio_session_hr_samples (
  session_started_at timestamptz not null references public.cardio_sessions(started_at) on delete cascade,
  sample_at timestamptz not null,
  bpm numeric not null check (bpm >= 0),
  primary key (session_started_at, sample_at)
);

create index cardio_session_hr_samples_session_idx
  on public.cardio_session_hr_samples (session_started_at);
```

Cascading delete keeps samples consistent if a session is ever removed (rare per PRD §7.11, but safe).

**Preprocessor changes** (`scripts/preprocess-health.py`):
- Stop discarding the raw HR sample stream after computing zone seconds.
- Emit a new `hr_samples: [{ ts, bpm }, ...]` array per session in the JSON output.
- Update `types/cardio.ts` `CardioSession` shape to include the optional `hr_samples` field; mirror in `lib/schemas/cardio.ts` and `scripts/lib/cardio-supabase.mjs` (the `KEEP IN SYNC WITH:` audit note already covers this).

**Import script** (`scripts/import-health.mjs` / `scripts/lib/cardio-supabase.mjs`):
- After upserting `cardio_sessions`, batch-insert HR samples into the new table.
- Batches of ~500 rows per request to stay under PostgREST's payload limits.
- Idempotency: delete-then-insert per session (samples are session-scoped; a re-import shouldn't leave stale samples behind).

**Detail page** (`app/training-facility/gym/session/[started_at]/page.tsx`):
- Server Component reads one session + its samples via a new `getCardioSession(startedAt)` in `lib/data/cardio-server.ts`.
- Renders: header (date, activity, duration), HR-curve line chart over time-in-session, zone breakdown (`<HrZoneBars />`), pace curve (running/walking only when sample-derived).
- Linked from the session log table — clicking the date cell navigates to the detail page. Composes naturally with the click-to-expand sibling (#164): inline preview in the table, "Open detail →" link to this page.

## Done when

- [ ] Migration adds `cardio_session_hr_samples` with FK + cascade
- [ ] `preprocess-health.py` emits `hr_samples` per session
- [ ] Import script upserts samples in batches; re-import is idempotent
- [ ] `getCardioSession(startedAt)` data-layer function returns session + samples
- [ ] `/training-facility/gym/session/[started_at]` route renders an HR curve + zone breakdown
- [ ] Session log rows link to the detail page
- [ ] Spot-check a real workout: HR curve matches Apple Health's view
- [ ] DB size after backfill stays well under 500 MB

## Out of scope

- Real-time HR streaming (manual export → import only)
- Per-stride / per-step running cadence samples (different data series — separate issue if/when needed)
- Editing samples in the UI (PRD §7.11 — fix in HealthKit, re-import)

## Dependencies

- #152 (Supabase-backed cardio data) — landed; this issue extends the same schema.
- Sibling issues #163 (zone strip) + #164 (click-to-expand) can ship in any order — independent work, no blocking dependency.

### PR DESCRIPTION
**Pull request by lucasklawrence**
## Summary

Adds `/training-facility/gym/session/[started_at]` — a per-session detail page that renders the full HR curve, zone breakdown, and metadata for one cardio workout. Closes the HR-zone trio: zone strip (#163) for scan, click-to-expand bars (#164) for focus, and now this page for drill-down.

Storage backing: a new `cardio_session_hr_samples` table (FK + cascade to `cardio_sessions`). The Python preprocessor now emits the raw HR sample stream alongside the existing zone aggregates; the import script delete-then-inserts samples per session in 500-row batches. Per the issue's storage estimate, ~50 MB/year of samples for a single user — Supabase free tier (500 MB) holds ~10 years.

## Changes

**Schema**
- `supabase/migrations/20260503120000_cardio_session_hr_samples.sql` — new table, applied to `court-vision` Supabase project.
- `(session_started_at, sample_at)` PK, FK with `on delete cascade`, `bpm >= 0` check, anon+authenticated SELECT RLS.

**Preprocessor (`scripts/preprocess-health.py`)**
- `aggregate_session_hr` now emits the in-window samples on `workout['hr_samples']` as `[{ts, bpm}, ...]`. Aggregate columns (avg/max/zones/efficiency) are unchanged.

**Schemas / types**
- `types/cardio.ts` — new `HrSample` interface, optional `hr_samples` on `CardioSession`.
- `lib/schemas/cardio.ts` — `CardioSessionHrSampleRowSchema`, `hrSampleRowToHrSample`, `hrSampleToRow`.
- `scripts/lib/cardio-supabase.mjs` — extends `CardioSessionSchema` to allow `hr_samples`.

**Import (`scripts/lib/cardio-supabase.mjs`)**
- New `upsertHrSamples` runs per-session delete-then-insert (500-row batches) after `cardio_sessions` upsert. Cascade FK handles session prune. Idempotent.
- `import-health.mjs` success log now reports the inserted-sample count.

**Data layer (`lib/data/cardio-server.ts`)**
- New `getCardioSession(startedAt)` returns `{session, samples}` or `null`. Uses the per-request server Supabase client.

**Page + view**
- `app/training-facility/gym/session/[started_at]/page.tsx` — Server Component, decodes the URL param, calls `getCardioSession`, `notFound()` on miss.
- `components/training-facility/gym/SessionDetailView.tsx` — header (date / activity / duration / distance / avg HR / max HR), HR-curve `<RoughLine>` over session-relative seconds, `<HrZoneBars>` zone breakdown. Empty-state for sessions without samples.

**Session-log links**
- Date cell in all four session-log tables (`AllCardioOverview` + `Stair` + `Treadmill` + `Track`) is now a `<Link>` that `stopPropagation`s so #164's row-expansion toggle stays intact.

## Test plan

- [ ] Re-run `npm run import-health -- <export.zip>` — success log mentions HR samples; Supabase shows rows in `cardio_session_hr_samples`.
- [ ] Click a date cell on `/training-facility/gym/overview` — navigates to the detail page; HR curve and zone bars render.
- [ ] On the detail page: spot-check the HR curve against Apple Health's view for the same workout.
- [ ] Try a session without HR samples (Apple Watch off) — page shows "No HR samples recorded" empty state, zone bars still show empty-chart state.
- [ ] Click on a row's date cell — navigates; clicking elsewhere on the row still toggles expansion (#164 behavior preserved).
- [ ] Repeat date-link checks on `/training-facility/gym/stair`, `/treadmill`, `/track`.
- [ ] Browser back from detail page → returns to the previous overview, scroll position preserved.
- [ ] Mobile (390×844): page header, stat cards, and chart cards stack cleanly; HR curve renders without overflow.
- [ ] Hand-typed bad URL like `/training-facility/gym/session/notarealtimestamp` — 404s rather than 500s.

## Notes

- Pace curve is out of scope for v1 per the upfront question pass — Apple Health's per-sample pace requires cadence data we don't currently parse. Defer to a follow-up issue.
- HR-curve x-axis is session-relative (`Mm SSs`), not wall-clock time — reads as "minute 8 of the workout."
- `[started_at]` route param is `encodeURIComponent`-encoded on the link, `decodeURIComponent`-decoded in the page — ISO timestamps contain `:` and `+`.
- Migration applied to `court-vision` Supabase project via MCP; SQL file is the canonical record.
- `npx tsc --noEmit` clean. 565 unit tests pass (10 new for `upsertHrSamples`). Build clean.

Closes #165.

## Summary by CodeRabbit

* **New Features**
  * Dedicated session detail pages with session metadata, interactive HR curve and time‑in‑zone charts.
  * Session dates are now clickable across training views to open session details.
  * Per-session raw HR sample support: ingestion, storage, and type/schema support.

* **Tests**
  * Added tests covering HR-sample upsert behavior, batching, and edge cases.

* **Chores**
  * Import logging now includes HR sample counts; DB migration adds HR-samples table.

### CODE REVIEW

File: app/training-facility/gym/session/[started_at]/page.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+export default async function TrainingFacilityGymSessionPage({
+  params,
+}: {
+  params: Promise<{ started_at: string }>
+}) {
+  if (!isTrainingFacilityEnabled()) notFound()
+
+  const { started_at: rawStartedAt } = await params
+  const startedAt = decodeURIComponent(rawStartedAt)
```
[the remark this answers is not in the document]
Comment (lucasklawrence) [reply app/training-facility/gym/session/[started_at]/page.tsx]: Fixed in 146a02b — `decodeURIComponent` is now wrapped in try/catch and routes to `notFound()` on `URIError`.

File: scripts/preprocess-health.py [hunk 2]
```diff
@@ -292,6 +299,16 @@ def aggregate_session_hr(workout: dict, samples: list[dict], max_hr: int) -> Non
...
 
+    # Emit the raw sample stream for the detail page. ISO format with
+    # offset matches `session.date`, so the TypeScript side parses both
+    # fields with the same code. Round BPM to one decimal — Apple Health
+    # emits whole integers but the JS schema accepts numeric, and the
+    # one-decimal slack costs nothing while leaving room for future
+    # smoothing without a re-import.
+    workout['hr_samples'] = [
+        {'ts': dt.isoformat(), 'bpm': round(bpm, 1)} for dt, bpm in in_window
+    ]
```
[the remark this answers is not in the document]
Comment (lucasklawrence) [reply scripts/preprocess-health.py:335]: Fixed in 146a02b — `aggregate_session_hr` now dedupes the in-window samples last-write-wins by timestamp before they reach the JSON output (and therefore the PK insert). Mirrors the existing `parse_simple_trend` pattern.

File: scripts/preprocess-health.py [hunk 3]
```diff
@@ -271,6 +278,22 @@ def aggregate_session_hr(workout: dict, samples: list[dict], max_hr: int) -> Non
...
+    for dt, bpm in in_window:
+        if deduped and deduped[-1][0] == dt:
+            deduped[-1] = (dt, bpm)
+        else:
+            deduped.append((dt, bpm))
+    in_window = deduped
+
     bpms = [bpm for _, bpm in in_window]
     workout['avg_hr'] = round(sum(bpms) / len(bpms), 1)
     workout['max_hr'] = round(max(bpms), 1)
```
[the remark this answers is not in the document]
Comment (lucasklawrence) [reply scripts/preprocess-health.py:308]: Fixed in 9b9ddd3 — `aggregate_session_hr` now filters `bpm >= 0` after the dedupe step (also catches NaN, since `nan >= 0` is False). Preserves the loud-fail behavior at the schema and DB layers as a backstop, but a single corrupt sample no longer aborts the whole import.

**Comment by lucasklawrence** [comment]
## Ready to merge

✅ `/review` clean on `9b9ddd3`
✅ CodeRabbit `SUCCESS` (last finding addressed; reviewer confirmed the fix is visible)
✅ Codex two findings from the initial review addressed in `146a02b`
✅ Vercel preview deployed
✅ e2e SUCCESS
✅ tsc clean, vitest 12/12 on `cardio-supabase.test.ts`

**Preview:** https://lucas-home-court-git-feat-165-h-a69336-lucasklawrences-projects.vercel.app

### Test plan

**Before re-importing data**

- [ ] `/training-facility/gym/overview` — date column on session-log rows now reads as a hyperlink; click navigates to `/training-facility/gym/session/[encoded-started_at]`.
- [ ] Detail page renders header (date, activity, duration, distance, avg/max HR), zone bars, and the **"No HR samples recorded"** placeholder where the curve would go (because the new table is empty until the import script runs).
- [ ] Same date-cell-as-link behavior on `/training-facility/gym/stair`, `/treadmill`, `/track`. Click-to-expand row toggle (#164) still works on the rest of the row.
- [ ] Visit `/training-facility/gym/session/%25` → 404 (malformed URL guard, codex P2).
- [ ] Visit `/training-facility/gym/session/2099-01-01T00:00:00Z` → 404 (no such session).
- [ ] Mobile (390×844): detail page lays out without horizontal scroll; chart cards fill width.

**After re-importing data**

- [ ] Run `npm run import-health -- <export.zip>` locally. Output now reports `N HR samples` count.
- [ ] Re-visit a detail page → HR curve renders end-to-end across the session window.
- [ ] Re-import the same export → idempotent; sample count unchanged; no DB errors.
- [ ] Spot-check one workout's curve against Apple Health's view for the same session — peaks/troughs should match.
- [ ] Inspect `cardio_session_hr_samples` row count in Supabase → matches "N HR samples" from the import log.

### Schema notes

- New migration `20260503120000_cardio_session_hr_samples.sql` was applied to the `court-vision` project as part of this PR.
- FK + cascade ensures samples clean up automatically when sessions are pruned.

### Merge

```
gh pr merge 170 --squash --delete-branch
```
